File: src/dotnet/Core/Application/Services/QueryParser.cs
using Rag.Candidates.Core.Domain.Configuration;
using Rag.Candidates.Core.Domain.Entities;
using Rag.Candidates.Core.Domain.Enums;

namespace Rag.Candidates.Core.Application.Services;

public sealed class QueryParser
{
    private readonly QueryParsingConfig _config;
    private readonly SeniorityMatcher _seniorityMatcher;
    private readonly TechnologyMatcher _technologyMatcher;
    private readonly ExperienceParser _experienceParser;

    public QueryParser(QueryParsingConfig? config = null)
    {
        _config = config ?? QueryParsingConfig.Default;
        _seniorityMatcher = new SeniorityMatcher(_config);
        _technologyMatcher = new TechnologyMatcher(_config);
        _experienceParser = new ExperienceParser();
    }

    public ParsedQuery Parse(string queryText)
    {
        var queryIntent = ExtractIntent(queryText);
        var requiredTechnologies = _technologyMatcher.ExtractTechnologies(queryText);
        var minSeniorityLevel = _seniorityMatcher.Match(queryText);
        var minYearsExperience = _experienceParser.ExtractYears(queryText);

        return new ParsedQuery
        {
            QueryText = queryText,
            QueryIntent = queryIntent,
            RequiredTechnologies = requiredTechnologies,
            MinSeniorityLevel = minSeniorityLevel,
            MinYearsExperience = minYearsExperience
        };
    }

    private QueryIntent ExtractIntent(string queryText)
    {
        var queryLower = queryText.ToLowerInvariant();

        foreach (var (intentKey, keywords) in _config.IntentKeywords)
        {
            foreach (var keyword in keywords)
            {
                if (queryLower.Contains(keyword.ToLowerInvariant()))
                {
                    return MapIntentKey(intentKey);
                }
            }
        }

        return QueryIntent.General;
    }

    private static QueryIntent MapIntentKey(string intentKey)
    {
        return intentKey switch
        {
            "find_best" => QueryIntent.FindBest,
            "list_all" => QueryIntent.ListAll,
            "compare" => QueryIntent.Compare,
            "explain" => QueryIntent.Explain,
            _ => QueryIntent.General
        };
    }
}

File: src/python/core/application/services/query_parser.py
from ...domain.entities.parsed_query import ParsedQuery
from ...domain.enums.query_intent import QueryIntent
from ...domain.configuration.query_parsing_config import (
    QueryParsingConfig,
    DEFAULT_QUERY_PARSING_CONFIG
)
from .seniority_matcher import SeniorityMatcher
from .technology_matcher import TechnologyMatcher
from .experience_parser import ExperienceParser


class QueryParser:
    def __init__(self, config: QueryParsingConfig = None):
        self._config = config or DEFAULT_QUERY_PARSING_CONFIG
        self._seniority_matcher = SeniorityMatcher(self._config)
        self._technology_matcher = TechnologyMatcher(self._config)
        self._experience_parser = ExperienceParser()
    
    def parse(self, query_text: str) -> ParsedQuery:
        query_intent = self._extract_intent(query_text)
        required_technologies = self._technology_matcher.extract_technologies(query_text)
        min_seniority_level = self._seniority_matcher.match(query_text)
        min_years_experience = self._experience_parser.extract_years(query_text)
        
        return ParsedQuery(
            query_text=query_text,
            query_intent=query_intent,
            required_technologies=required_technologies,
            min_seniority_level=min_seniority_level,
            min_years_experience=min_years_experience
        )
    
    def _extract_intent(self, query_text: str) -> QueryIntent:
        query_lower = query_text.lower()
        
        for intent_key, keywords in self._config.intent_keywords.items():
            for keyword in keywords:
                if keyword.lower() in query_lower:
                    return self._map_intent_key(intent_key)
        
        return QueryIntent.GENERAL
    
    def _map_intent_key(self, intent_key: str) -> QueryIntent:
        mapping = {
            "find_best": QueryIntent.FIND_BEST,
            "list_all": QueryIntent.LIST_ALL,
            "compare": QueryIntent.COMPARE,
            "explain": QueryIntent.EXPLAIN,
        }
        return mapping.get(intent_key, QueryIntent.GENERAL)
